internal/mesh/internal/controllers/sidecarproxy/builder/local_app.go

```diff
@@ -0,0 +1,141 @@
+package builder
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/envoyextensions/xdscommon"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1/pbproxystate"
+)
+
+func (b *Builder) BuildLocalApp(workload *pbcatalog.Workload) *Builder {
+	return b.addInboundListener(xdscommon.PublicListenerName, workload).
+		addInboundRouters(workload).
+		addInboundTLS()
+}
+
+func (b *Builder) getLastBuiltListener() *pbproxystate.Listener {
+	lastBuiltIndex := len(b.proxyStateTemplate.ProxyState.Listeners) - 1
+	return b.proxyStateTemplate.ProxyState.Listeners[lastBuiltIndex]
+}
+
+func (b *Builder) addInboundListener(name string, workload *pbcatalog.Workload) *Builder {
+	listener := &pbproxystate.Listener{
+		Name:      name,
+		Direction: pbproxystate.Direction_DIRECTION_INBOUND,
+	}
+
+	// We will take listener bind port from the workload for now.
+	// Find mesh port.
+	var meshPort string
+	for portName, port := range workload.Ports {
```

Comment:
(aside)
Perhaps we should add a helper method to `proto-public/pbcatalog/v1alpha1/workload_addon.go` with a `Workload` receiver called `GetMeshPort()` that does this trick? I feel like I personally have written something like this a bunch, and have seen this a few times in various forms in this PR.